src/xmerl_c14n.erl

```diff
@@ -0,0 +1,527 @@
+        [] -> ["?>", NameString, "<?" | Acc];
+        _ -> ["?>", Value, " ", NameString, "<?" | Acc]
+    end;
+c14n(#xmlDocument{content = Kids}, KnownNS, ActiveNS, Comments, InclNs, Acc) ->
+    case lists:foldl(fun(Kid, AccIn) ->
+                             case c14n(Kid, KnownNS, ActiveNS, Comments, InclNs, AccIn) of
+                                 AccIn -> AccIn;
+                                 Other -> ["\n" | Other]
+                             end
+                     end, Acc, Kids) of
+        ["\n" | Rest] -> Rest;
+        Other -> Other
+    end;
+c14n(#xmlAttribute{nsinfo = NsInfo, name = Name, value = Value}, _KnownNs, ActiveNS, _Comments, _InclNs, Acc) ->
+    case NsInfo of
+        {Ns, NName} ->
+            case lists:member(Ns, ActiveNS) of
+                true -> ["\"",xml_safe_string(Value, true),"=\"",NName,":",Ns," " | Acc];
+                _ -> error("attribute namespace is not active")
+            end;
+        _ ->
+            ["\"",xml_safe_string(Value, true),"=\"",atom_to_list(Name)," " | Acc]
+    end;
+c14n(Elem = #xmlElement{}, KnownNSIn, ActiveNSIn, Comments, InclNs, Acc) ->
+    Namespace = Elem#xmlElement.namespace,
+    Default = case Elem#xmlElement.nsinfo of
+                  [] -> Namespace#xmlNamespace.default;
+                  _ -> [] % omit a default namespace if it is not visibly utilized.
+              end,
+    {ActiveNS, ParentDefault} = case ActiveNSIn of
+                                    [{default, P} | Rest] -> {Rest, P};
+                                    Other -> {Other, ''}
+                                end,
+    %% add any new namespaces this element has that we haven't seen before
+    KnownNS = lists:foldl(fun({Ns, Uri}, Nss) ->
+                                  case proplists:is_defined(Ns, Nss) of
+                                      true -> Nss;
+                                      _ -> [{Ns, atom_to_list(Uri)} | Nss]
+                                  end
+                          end, KnownNSIn, Namespace#xmlNamespace.nodes),
+
+    %% now figure out the minimum set of namespaces we need at this level
+    NeededNs = needed_ns(Elem, InclNs),
+    %% and all of the attributes that aren't xmlns
+    Attrs = clean_sort_attrs(Elem#xmlElement.attributes),
+
+    %% we need to append any xmlns: that our parent didn't have (ie, aren't in ActiveNS) but
+    %% that we need
+    NewNS = (InclNs -- NeededNs) ++ NeededNs -- ActiveNS,
+    NewActiveNS = ActiveNS ++ NewNS,
+
+    %% the opening tag
+    Acc1 = case Elem#xmlElement.nsinfo of
+               {ENs, EName} ->
+                   [EName, ":", ENs, "<" | Acc];
+               _ ->
+                   [atom_to_list(Elem#xmlElement.name), "<" | Acc]
+           end,
+    %% xmlns definitions
+    {Acc2, FinalActiveNS} =
+        if
+            not (Default =:= []) andalso not (Default =:= ParentDefault) ->
+                {["\"", xml_safe_string(Default, true), " xmlns=\"" | Acc1],
```

Comment:
Can we not just always escape quotes with `&quot;` and not have to use the boolean second argument?

Reply:
I didn't touch any part of this code so I can't comment on it.